### ISSUE TITLE
weird behaviour when tampering state file

### ISSUE THREAD

**Issue by notdodo** [pulumi/pulumi#11969]
### What happened?

During a security threat analysis I've stumbled on the tampering of a Pulumi state file and I've found out that duplicating a resource in the state file can cause a remote resource be deleted.

### Steps to reproduce

Run a Kubernetes cluster using Pulumi; create resources and save the Pulumi state in a local file (for testing purposes).

- `pulumi logout`
- `pulumi login file:///home/notdodo/pulumi-k8s/`

Edit the file `./.pulumi/stacks/pulumi-k8s.json` (`pulumi-k8s` is just the stack name).

Append to the list of resources another one with a different URN; let's refer to it a the FAKE resource. For example add the following namespace resource:

```json
{
   "urn":"urn:pulumi:pulumi-k8s::pulumi-k8s::kubernetes:core/v1:Namespace::test-ns0",
   "custom":true,
   "id":"test-ns-49cf8b8f",
   "type":"kubernetes:core/v1:Namespace",
   "inputs":{
      "apiVersion":"v1",
      "kind":"Namespace",
      "metadata":{
         "annotations":{
            "pulumi.com/autonamed":"true"
         },
         "labels":{
            "app.kubernetes.io/managed-by":"pulumi"
         },
         "name":"test-ns-49cf8b8f"
      }
   },
   "parent":"urn:pulumi:pulumi-k8s::pulumi-k8s::pulumi:pulumi:Stack::pulumi-k8s-pulumi-k8s",
   "provider":"urn:pulumi:pulumi-k8s::pulumi-k8s::pulumi:providers:kubernetes::default_3_24_0_alpha_1674143625::c86be6c1-42c5-492e-b3d3-fd4fff8b56f9",
   "propertyDependencies":{
      "apiVersion":null,
      "kind":null
   }
}
```
Actually the namespace `test-ns` esists and it's defined in another identical resource with URN `urn:pulumi:pulumi-k8s::pulumi-k8s::kubernetes:core/v1:Namespace::test-ns0` (let's refer to it as the REAL resource); all other fields are identical to the REAL `test-ns` resource.
The `test-ns` is defined in the Pulumi python code and it's deployed on the cluster.

Now Run:

- `pulumi refresh`: the FAKE resource (`test-ns0`) will be updated; the REAL resource is not reported in the preview

![image](https://user-images.githubusercontent.com/6991986/214311430-4b62cfe5-9f6c-47e8-9b3c-e3e658418097.png)

Now Run:

- `pulumi up`: the FAKE resource (`test-ns0`) will be deleted

![image](https://user-images.githubusercontent.com/6991986/214311541-16724ebd-2c28-4f4b-901d-c3b814a37c75.png)

During the deploy actually the REAL resource will be deleted from the Kubernetes cluster:

![image](https://user-images.githubusercontent.com/6991986/214311585-0bc7da5f-2c30-4763-b65b-c05585ea6083.png)

That’s because the FAKE resource refers to a valid namespace name.

Further inspection of the Pulumi state:

1. the REAL resource is still present in the code and in the state; a `pulumi refresh` will remove it only from the state file
2. even if the repository code still declares that the resource should be create, Pulumi will not create again the resource until another `pulumi refresh` and `pulumi up`.

Using the same _technique_ if the [pulumi-command provider](https://www.pulumi.com/registry/packages/command/) is available it's also possible to execute commands on the device/pipeline running/managing the pulumi stack.

For example a local command execution the resource looks like:

```json
{
   "urn":"urn:pulumi:pulumi-k8s::pulumi-k8s::kubernetes:helm.sh/v3:Release$command:local:Command::unseal-vault",
   "custom":true,
   "id":"unseal-vaultc1ff9ffb",
   "type":"command:local:Command",
   "inputs":{
      "create":"./vault_utils/unseal.sh"
   },
   "outputs":{
      "create":"./vault_utils/unseal.sh",
      "stderr":"",
      "stdout":"Not Sealed"
   },
   "parent":"urn:pulumi:pulumi-k8s::pulumi-k8s::kubernetes:helm.sh/v3:Release::vault",
   "dependencies":[
      "urn:pulumi:pulumi-k8s::pulumi-k8s::kubernetes:helm.sh/v3:Release::vault"
   ],
   "provider":"urn:pulumi:pulumi-k8s::pulumi-k8s::pulumi:providers:command::default_0_7_1_alpha_1674088865::accb3d69-eb46-4295-8340-e432130739f2",
   "propertyDependencies":{
      "create":null
   }
}
```

In the `create` fields there is the command to execute, you can also execute inline bash commands for example.
In this example the provider is used to unseal a vault deployment but it’s just an example and only for testing purposes.

Let’s create another tampered `unseal-vault0` resource that executes a command also for `delete` and `update` triggers.

NOTE: the provider of Pulumi Command must be present.

```json
{
   "urn":"urn:pulumi:pulumi-k8s::pulumi-k8s::kubernetes:helm.sh/v3:Release$command:local:Command::unseal-vault0",
   "custom":true,
   "id":"unseal-vault76838b44",
   "type":"command:local:Command",
   "inputs":{
      "create":"touch /tmp/pwned",
      "delete":"touch /tmp/pwned",
      "update":"touch /tmp/pwned"
   },
   "outputs":{
      "create":"touch /tmp/pwned",
      "delete":"touch /tmp/pwned",
      "stderr":"",
      "stdout":"Not Sealed",
      "update":"touch /tmp/pwned"
   },
   "parent":"urn:pulumi:pulumi-k8s::pulumi-k8s::kubernetes:helm.sh/v3:Release::vault",
   "dependencies":[
      "urn:pulumi:pulumi-k8s::pulumi-k8s::kubernetes:helm.sh/v3:Release::vault"
   ],
   "provider":"urn:pulumi:pulumi-k8s::pulumi-k8s::pulumi:providers:command::default_0_7_1_alpha_1674088865::accb3d69-eb46-4295-8340-e432130739f2",
   "propertyDependencies":{
      "create":null
   }
}
```

Here we’ll try to execute the command `touch /tmp/pwned`

NOTE: the REAL `unseal-vault` resource is still present and no changes to code are performed.

Run:

- `pulumi up`: the new `unseal-vault0`resource will be deleted

![image](https://user-images.githubusercontent.com/6991986/214313970-d4e59da0-641a-494c-b229-14673dd7ab4c.png)

But the result of the update triggers the execution of the command:

![image](https://user-images.githubusercontent.com/6991986/214313999-10a4bf91-81d6-44a1-9f26-0aa51026c47c.png)

### Expected Behavior

Not code-defined resource is deleted from the state; actual resource deployed is not deleted.

### Actual Behavior

Basically it's possible to arbitrary delete any resource without actual access to infrastructure or it's possible to run arbitrary commands on the pipeline or admin devices running/managing the Pulumi stack.
It seems that Pulumi is not handling correctly the resources.

### Output of `pulumi about`

```
CLI          
Version      3.52.1
Go Version   go1.19.5
Go Compiler  gc

Plugins
NAME                     VERSION
command                  0.7.1-alpha.1674088865
kubernetes               3.24.0-alpha.1674143625
kubernetes-cert-manager  0.0.5
oci                      0.5.0
python                   3.10

Host     
OS       ubuntu
Version  22.04
Arch     x86_64
```

**Comment by RobbieMcKinstry** [comment]
Hi @notdodo , I appreciate your thorough write-up of this issue.

RE: the filestate backend. The filestate backend gives total local control over the Pulumi state – because the file is saved locally, anyone with `-rw` permission to the file can modify the state. It's inherently error-prone to edit the state directly. More generally, editing the state with a `pulumi state` subcommand is dangerous, so we've included a confirmation message to ensure users are aware of this. In general, state edits are intended for "surgical" remediations when something goes wrong _e.g._, the Pulumi process is killed during execution and left in a bad state, or a cloud provider has a bug in the implementation of their provider. Administrators should ensure state access adheres to the Principle of Least Privilege.

From the point of view of the Pulumi engine, it can only really act with the information it's given. It has to assume the state file is trusted. If someone has tampered with the state file, then the engine can't differentiate the tampered resources from a non-tampered resource -- there's no way for the engine to know if an edit to the statefile was made in good faith during error recovery or by a malicious user.

That's an interesting point you've raised about using the `command` provider to inject an attack against the running program. When running a Pulumi program, you're executing real code in a general-purpose language. If a malicious user were to inject arbitrary code into a company's Pulumi program, and the company decides to execute that program, then they've executing arbitrary source – that's also unavoidable. Most organizations protect against this using the mechanisms provided by their source control provider (_e.g._ GitHub: thorough PR review, SSO, projecting branches from force-push, and requiring status checks before merge), or via [Pulumi policies](https://www.pulumi.com/docs/guides/crossguard/).

IMO Pulumi can either choose to be flexible in the face of errors or secure under all circumstances. Ideally, running programs would always be able to save intermediate state when given a KILL signal, and provider implementation would never least state in a buggy spot. If that were the case, we could eliminate 99.9% of the usage of `pulumi cancel` command. Unfortunately, I don't think it's realistic for Pulumi ever to reach this spot, which means we will always need to support `pulumi state` subcommands, and deal with the fallout that entails (in this case the fallout is an increased security burden on the user to ensure their state files are protected).

I appreciate your feedback. It's been helpful to me in thinking about the attack surface of Pulumi programs. However, I'm going to make this issue as "by-design". I believe ultimately it must fall to the user to control access to the Pulumi statefile as part of their threat model.

**Comment by notdodo** [comment]
Hi @RobbieMcKinstry! Thank you for the explanation.

> The filestate backend gives total local control over the Pulumi state – because the file is saved locally, anyone with `-rw` permission to the file can modify the state.

Yes! that's the whole point of the analysis; let say that for compliance/legal/control/whatever you can't use the Pulumi _cloud_ to store the statefile, this needs to be protected in both read and write modes (as you said "Administrators should ensure state access adheres to the Principle of Least Privilege"). But I'm not talking about the usage of `pulumi state` but simply editing the file with the purpose to simulate _what can happen if_; nor I want to propose to remove features and such.

My point of view is that the behaviour is somewhat strange that can be associated a bug of the engine.
When manually adding a new resource to the state my expectation where:
1. pulumi tries to actually create the resource on kubernetes; then the creation fails since the namespace exists
2. an error on pulumi
3. the additional resource is deleted since on the code there no such thing or there is but is yet present (I created a doppler in the statefile).

What I hardly believe is that, in case of another resource with the same properties (and not present in the code, which in my opinion is the source of truth), the result is the deletion of the actual deployed resource is a "by design" choice and not a bug in the engine.

> It has to assume the state file is trusted.

If you think that this is _intended_ and not a bug I think that this sentence can recap all our efforts to write this comments :smile:

**Comment by Frassle** [comment]
> (and not present in the code, which in my opinion is the source of truth)

I think this is where your thinking is off. As far as the engine is concerned both the program _and_ the state file are the source of truth. The program alone is insufficient to describe what operations to do.
